Add config option to create to control whether a new option is selected or not

fgs-dbudwin opened the pull request:
If the creation of the new option in production code is dependent on some other service--like inserting the new option in a database--there is a potential for that operation to fail.  When writing a unit test to verify how that failure is handled, it may not be desirable to select the new option given to `create` if the insert into the database failed.

This commit introduces a new option in the `CreateConfig` object called `autoSelect` that will enable or disable the automatic selection of the new option.

See issue #35 for more details.

romgain commented on package.json:
```diff
@@ -1,6 +1,6 @@
 {
   "name": "react-select-event",
-  "version": "4.1.4",
+  "version": "4.2.0",
```
Would you mind reverting this?
It'll get done when I publish the package :slightly_smiling_face:

romgain commented on src/__tests__/select-event.test.tsx:
```diff
@@ -124,13 +127,21 @@ describe("The select event helpers", () => {
     expect(form).toHaveFormValues({ food: "" });
 
     await selectEvent.create(input, "papaya", {
+      autoSelect: true, 
```
Since `autoSelect` already defaults to true, can we remove this line?

fgs-dbudwin replied:
What I think is going on here, is that `config` is being overwritten and the value for `config` only sets `createOptionText`.  A more longform way of thinking of this:

```
const config: CreateConfig = {
  createOptionText: "Add new option",
}

console.log(JSON.stringify(config));

await selectEvent.create(input, "papaya", config);
```

This will print the following `{"createOptionText":"Add new option"}`...there is no `autoSelect` key in that object.  If I remove `autoSelect: true`, the test will fail.  So `autoSelect` defaults to `true`, unless you provide a value for `config`, then `autoSelect` will have whatever value the `config` parameter has for `autoSelect`.

Disclaimer: I'm not a JS/TS expert, so I may simply be doing something wrong 😁

romgain replied:
Oh I see!

The syntax is pretty weird, but to make the `autoSelect` option really optional, you can write line 104 like this:

```ts
   { autoSelect = true, ...config }: CreateConfig = {} 
```

(and then directly use `autoSelect` instead of `config.autoSelect` line 112)

fgs-dbudwin replied:
Make sense, learned something new today!

romgain commented on src/__tests__/select-event.test.tsx:
```diff
@@ -146,7 +157,7 @@ describe("The select event helpers", () => {
     expect(form).toHaveFormValues({ food: "" });
 
-    await selectEvent.create(input, "papaya", { createOptionText: "papaya" });
+    await selectEvent.create(input, "papaya", { autoSelect: true, createOptionText: "papaya" });
```
Same here - can we remove all the `autoSelect: true`?

fgs-dbudwin replied:
Ditto.

fgs-dbudwin commented:
I believe I have addressed the feedback and pushed a fixup! commit.  If you find the results satisfactory, I will squash the fixup! commit and re-push.

fgs-dbudwin commented:
Excellent, I squashed my fixup commit to tidy the commit history a bit!  Thanks for maintaining this library, it's extremely useful!

fgs-dbudwin commented:
Sigh, I think I may have done this slightly wrong...I should have stuck with what I said in issue #35.

It should look like this:

```js
await select(input, createOptionText, config);
  
if (autoSelect) {
  await findByText(getReactSelectContainerFromInput(input), option);
}
```

If I understand the code correctly, that `await select(...)` is mimicking the user clicking `Create: "[whatever they entered]`".  I want that interaction of clicking the create to happen, but when the create fails on the backend, I don't want this library to then try and find the new value (`findByText`) in the select menu because it shouldn't exist.

I fixed it locally, but the test are being tricky...

romgain commented:
The best way would probably be to add a minimal reproduction of what you actually need in your production code to the tests.
That would document the intent well, and will make sure that we don't break this functionality in the future :slightly_smiling_face: 

I can probably help with the test if you share a minimal component that does what you need

romgain commented:
Right - I think I see what you want.

Can we add an option that controls `await findByText(getReactSelectContainerFromInput(input), option);` instead of just removing it though? There's not reason for this to be a breaking change
